### ISSUE TITLE
ZFS zpools missing after running latest round of updates on CentOS 7.2.

### ISSUE THREAD

**Issue by vmorgo** [zfsonlinux/zfs#5171]
I thought I'd better bring this to your attention--it's a rather inconvenient (but fortunately fairly easy to work around) bug that has been introduced either with the latest version of ZOL or the latest round of CentOS/RedHat updates:

The latest round of updates for CentOS 7.2/Redhat 7.2 has broken the ZFS-On-Linux ability to automount any zpools upon system boot-up.  

It is now necessary to run "zpool import -f <poolname> upon _every_ boot of the affected server.

One work-around is to include this in some system startup script

On CentOS 7.2, there are no files such as:

/etc/init/zpool-import.conf

As the CentOS 7.2 system apparently no longer uses /etc/init (in favor of Systemctl??), etc/init/ no longer exists.

or 

/etc/default/zfs

This was tried with no change.

The following debugging was attempted:

[root@kalypso ~]# zpool status
no pools available

[root@kalypso ~]# yum update zfs*
Loaded plugins: fastestmirror, langpacks
Loading mirror speeds from cached hostfile
- base: mirror.atlanticmetro.net
- elrepo: mirrors.evowise.com
- epel: mirror.cc.columbia.edu
- extras: mirror.metrocast.net
- updates: mirror.atlanticmetro.net
  No packages marked for update

[root@kalypso ~]# dmesg | grep "using hostid"
[root@kalypso ~]# dmesg | grep "using hostid"
[root@kalypso ~]# 

[root@kalypso ~]# zpool import -f virtpool
[root@kalypso ~]# zdb 
virtpool:
    version: 5000
    name: 'virtpool'
    state: 0
    txg: 3061820
    pool_guid: 18233218770433194016
    errata: 0
    hostname: 'kalypso.bumc.bu.edu'
    vdev_children: 1
    vdev_tree:
        type: 'root'
        id: 0
        guid: 18233218770433194016
        children[0]:
            type: 'mirror'
            id: 0
            guid: 13429003678131489853
            metaslab_array: 34
            metaslab_shift: 33
            ashift: 9
            asize: 1000189984768
            is_log: 0
            create_txg: 4
            children[0]:
                type: 'disk'
                id: 0
                guid: 13075379023475076914
                path: '/dev/sdb1'
                whole_disk: 1
                DTL: 237
                create_txg: 4
            children[1]:
                type: 'disk'
                id: 1
                guid: 3638277088579218112
                path: '/dev/sdc1'
                whole_disk: 1
                DTL: 236
                create_txg: 4
            children[2]:
                type: 'disk'
                id: 2
                guid: 6848820760391464338
                path: '/dev/sdd1'
                whole_disk: 1
                DTL: 231
                create_txg: 4
    features_for_read:
        com.delphix:hole_birth
        com.delphix:embedded_data
[root@kalypso ~]# zdb | grep hostid
[root@kalypso ~]# zdb | grep `hostid`
[root@kalypso ~]# 

ZFS versions in use:
[root@phoenix ~]# rpm -q -a | grep zfs
zfs-dkms-0.6.5.8-1.el7.centos.noarch
libzfs2-0.6.5.8-1.el7.centos.x86_64
zfs-release-1-3.el7.centos.noarch
zfs-0.6.5.8-1.el7.centos.x86_64

Present work-around is to either manually mount the volumes after booting the host, to avoid the current upgrade to zfs-0.6.5.8-1.el7.centos.x86_64, or to add the line 

zpool import -f <poolname>, <poolname>.... 

to the auto-start script of your choice or just run it manually.  In my case, I added it to /etc/rc.d/rc.local _specifically_ because I am not aware of ZFS requiring Systemctl to stop, start, etc., it and, and it is easy enough to chmod +/- x to enable rc.local as needed (and, since /etc/rc.d/rc.local is just about the last thing executed during boot-up, ZFS has plenty of time to get its act together.

NOTE:  Please reference Howto #1839 in any response.  This will enable me to update my Howtos database.

Many thanks!

**Comment by deajan** [comment]
First, of all, your howto number seems incorrect and leads to 3 years old dedup PR.

Then, why would you use initV style startup scripts on CentOS 7 when the whole system uses systemd ?

A simple command to enable the auto importing on startup is
```
systemctll enable zfs-import-cache
```

or the following if you prefer scanning each time instead of using cache
```
systemctl enable zfs-import-scan
```

**Comment by vmorgo** [comment]
Will try that. Question then becomes 'why has systemctl enable zfs-import-cache been omitted or disabled withe the current version of ZOL when it was apparently included in the update process previous versions?'. One has to wonder at such omissions. Many newer users may not think or know where to look when, for no apparent reason they update and suddenly find no zfs pools are available. Sadly this omission is undocumented and the consequences could. Be dire indeed. The howto reference is for my internal howtos database, not the ZFS On Linux one. Sorry if I wasn't clear on that. Thank you for your help with this,  Osiris                                                                                                                                                                         Sent from my BlackBerry - the most secure mobile device                                                                                                                                                                                                                   From: notifications@github.comSent: September 27, 2016 03:24To: zfs@noreply.github.comReply-to: reply@reply.github.comCc: lafeyette.management@gmail.com; author@noreply.github.comSubject: Re: [zfsonlinux/zfs] ZFS zpools missing after running latest round of updates on CentOS 7.2. (#5171)  First, of all, your howto number seems incorrect and leads to 3 years old dedup PR.

Then, why would you use initV style startup scripts on CentOS 7 when the whole system uses systemd ?

A simple command to enable the auto importing on startup is

systemctll enable zfs-import-cache


or the following if you prefer scanning each time instead of using cache

systemctl enable zfs-import-scan


—You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by vmorgo** [comment]
Please disregard.  The problem reported yesterday seems to occur when the following happens:

1.  Centos 7.2 is used with a non-CentOS kernel, e.g., one of the ElRepo ML or LT series kernels such as 4.4.16LT or 4.6ML.
2.  The problem is that, for whatever reason, systemctl enable zfs-import-cache does not happen during the install or upgrade to ZFS to 0.6.5.8-1.el7.  Running systemctl enable zfs-import-cache after the install works just fine.  When rebooted, the machine with the ElRepo ML or LT series kernel will boot and start its ZPOOLS properly.

Thanks to deajan for his help with this.  I guess it's all part of the learning experience!

NOTE:  I am not 100% sure it will always work with the stock CentOS kernel as I have only one macheine that runs the stock kernel and I have it locked at a relatively old version due to previous CentOS kernel updates "breaking" ZFS.  All other machines I own/manage run the newer ElRepo 4.6ML or the ElRepo 4.4.16LT kernel.  These newer kernels are (almost always) a lot less troublesome.

A./

**Comment by vmorgo** [comment]
However, it is definitely worth mentioning that there is a problem with systemctl enable zfs-import-cache with versions of zfs-on-linux earlier than 0.6.5.8-1.el7, namely that the command WILL FAIL with the following message:

[root@knossos ~]# systemctl enable zfs-import-cache
The unit files have no [Install] section. They are not meant to be enabled
using systemctl.
Possible reasons for having this kind of units are:
1) A unit may be statically enabled by being symlinked from another unit's
   .wants/ or .requires/ directory.
2) A unit's purpose may be to act as a helper for some other unit which has
   a requirement dependency on it.
3) A unit may be started when needed via activation (socket, path, timer,
   D-Bus, udev, scripted systemctl call, ...).


Evidently, one of the items included with the upgrade to 0.6.5.8-1.el7 is that the unit files for zfs-import-cache (and zfs-import-cache) have been added or modified to include an "Install" section and that these unit files either did not exist or were not available to systemctl in previous versions.

This probably should be somehow documented as others using non-stock CentOS kernels are likely to be tripped up by this change in the unit files for zfs-import-cache and zfs-import-scan.
